File: machine/ts/machine/devices/gx/system_vram_port.ts
import {
	IO_GX_GPU_SYSTEM_VRAM_CONTROL,
	IO_GX_GPU_SYSTEM_VRAM_DATA,
	IO_GX_GPU_SYSTEM_VRAM_POSITION,
	IO_GX_GPU_SYSTEM_VRAM_SIZE,
	IO_GX_GPU_SYSTEM_VRAM_STATUS,
} from '../../bus/io';
import type { Value } from '../../cpu/cpu';
import { Memory } from '../../memory/memory';

export const GX_GPU_SYSTEM_VRAM_X = 512;
export const GX_GPU_SYSTEM_VRAM_Y = 0;
export const GX_GPU_SYSTEM_VRAM_WIDTH = 256;
export const GX_GPU_SYSTEM_VRAM_HEIGHT = 256;
export const GX_GPU_SYSTEM_VRAM_PORT_COMMAND_CAPACITY = 256;
export const GX_GPU_SYSTEM_VRAM_PORT_WORD_CAPACITY = 0x8000;
export const GX_GPU_SYSTEM_VRAM_PORT_CONTROL_START = 1 << 0;
export const GX_GPU_SYSTEM_VRAM_PORT_CONTROL_RESET = 1 << 1;
export const GX_GPU_SYSTEM_VRAM_PORT_STATUS_BUSY = 1 << 0;
export const GX_GPU_SYSTEM_VRAM_PORT_STATUS_WRITE_READY = 1 << 1;
export const GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW = 1 << 2;
export const GX_GPU_SYSTEM_VRAM_PORT_STATUS_PENDING = 1 << 3;
export const GX_GPU_SYSTEM_VRAM_PORT_STATUS_REMAINING_SHIFT = 8;

const gxGpuSystemVramUnmappedReadError = new Error('GX system VRAM port read handler received an unmapped register.');
const gxGpuSystemVramUnmappedWriteError = new Error('GX system VRAM port write handler received an unmapped register.');

export function gxGpuSystemVramX(positionWord: number): number {
	return GX_GPU_SYSTEM_VRAM_X + (positionWord & (GX_GPU_SYSTEM_VRAM_WIDTH - 1));
}

export function gxGpuSystemVramY(positionWord: number): number {
	return GX_GPU_SYSTEM_VRAM_Y + ((positionWord >>> 16) & (GX_GPU_SYSTEM_VRAM_HEIGHT - 1));
}

export function gxGpuSystemVramColumnX(positionWord: number, column: number): number {
	return GX_GPU_SYSTEM_VRAM_X + ((positionWord + column) & (GX_GPU_SYSTEM_VRAM_WIDTH - 1));
}

export function gxGpuSystemVramRowY(positionWord: number, row: number): number {
	return GX_GPU_SYSTEM_VRAM_Y + (((positionWord >>> 16) + row) & (GX_GPU_SYSTEM_VRAM_HEIGHT - 1));
}

export function gxGpuSystemVramWidth(sizeWord: number): number {
	return (((sizeWord & 0xff) - 1) & 0xff) + 1;
}

export function gxGpuSystemVramHeight(sizeWord: number): number {
	return ((((sizeWord >>> 16) & 0xff) - 1) & 0xff) + 1;
}

export type GxGpuSystemVramPortState = {
	positionWord: number;
	sizeWord: number;
	controlWord: number;
	dataWord: number;
	statusWord: number;
	commandCount: number;
	presentCommandCount: number;
	wordCount: number;
	activePositionWord: number;
	activeSizeWord: number;
	activeWordStart: number;
	activeWordsRemaining: number;
	commandPositionWord: number[];
	commandSizeWord: number[];
	commandWordStart: number[];
	words: number[];
};

export type GxGpuSystemVramPortView = Readonly<{
	serial: number;
	commandCount: number;
	presentCommandCount: number;
	wordCount: number;
	commandPositionWord: ArrayLike<number>;
	commandSizeWord: ArrayLike<number>;
	commandWordStart: ArrayLike<number>;
	words: ArrayLike<number>;
}>;

let gxGpuSystemVramNextSerial = 0;

export class GxGpuSystemVramPort implements GxGpuSystemVramPortView {
	public serial = 0;
	public commandCount = 0;
	public presentCommandCount = 0;
	public wordCount = 0;
	public readonly commandPositionWord = new Uint32Array(GX_GPU_SYSTEM_VRAM_PORT_COMMAND_CAPACITY);
	public readonly commandSizeWord = new Uint32Array(GX_GPU_SYSTEM_VRAM_PORT_COMMAND_CAPACITY);
	public readonly commandWordStart = new Uint32Array(GX_GPU_SYSTEM_VRAM_PORT_COMMAND_CAPACITY);
	public readonly words = new Uint32Array(GX_GPU_SYSTEM_VRAM_PORT_WORD_CAPACITY);

	private positionWord = 0;
	private sizeWord = 0;
	private controlWord = 0;
	private dataWord = 0;
	private statusWord = 0;
	private activePositionWord = 0;
	private activeSizeWord = 0;
	private activeWordStart = 0;
	private activeWordsRemaining = 0;

	public constructor(memory: Memory) {
		memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_POSITION, this, GxGpuSystemVramPort.readRegister);
		memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_SIZE, this, GxGpuSystemVramPort.readRegister);
		memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_CONTROL, this, GxGpuSystemVramPort.readRegister);
		memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_DATA, this, GxGpuSystemVramPort.readRegister);
		memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_STATUS, this, GxGpuSystemVramPort.readRegister);
		memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_POSITION, this, GxGpuSystemVramPort.writeRegister);
		memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_SIZE, this, GxGpuSystemVramPort.writeRegister);
		memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_CONTROL, this, GxGpuSystemVramPort.writeRegister);
		memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_DATA, this, GxGpuSystemVramPort.writeRegister);
	}

	public reset(): void {
		this.positionWord = 0;
		this.sizeWord = 0;
		this.controlWord = 0;
		this.dataWord = 0;
		this.statusWord = 0;
		this.commandCount = 0;
		this.presentCommandCount = 0;
		this.wordCount = 0;
		this.activePositionWord = 0;
		this.activeSizeWord = 0;
		this.activeWordStart = 0;
		this.activeWordsRemaining = 0;
		this.publishRevision();
	}

	public captureState(): GxGpuSystemVramPortState {
		return {
			positionWord: this.positionWord,
			sizeWord: this.sizeWord,
			controlWord: this.controlWord,
			dataWord: this.dataWord,
			statusWord: this.statusWord,
			commandCount: this.commandCount,
			presentCommandCount: this.presentCommandCount,
			wordCount: this.wordCount,
			activePositionWord: this.activePositionWord,
			activeSizeWord: this.activeSizeWord,
			activeWordStart: this.activeWordStart,
			activeWordsRemaining: this.activeWordsRemaining,
			commandPositionWord: Array.from(this.commandPositionWord.subarray(0, this.commandCount)),
			commandSizeWord: Array.from(this.commandSizeWord.subarray(0, this.commandCount)),
			commandWordStart: Array.from(this.commandWordStart.subarray(0, this.commandCount)),
			words: Array.from(this.words.subarray(0, this.wordCount)),
		};
	}

	public restoreState(state: GxGpuSystemVramPortState): void {
		this.positionWord = state.positionWord >>> 0;
		this.sizeWord = state.sizeWord >>> 0;
		this.controlWord = state.controlWord >>> 0;
		this.dataWord = state.dataWord >>> 0;
		this.statusWord = state.statusWord >>> 0;
		this.commandCount = state.commandCount;
		this.presentCommandCount = state.presentCommandCount;
		this.wordCount = state.wordCount;
		this.activePositionWord = state.activePositionWord >>> 0;
		this.activeSizeWord = state.activeSizeWord >>> 0;
		this.activeWordStart = state.activeWordStart;
		this.activeWordsRemaining = state.activeWordsRemaining;
		this.commandPositionWord.set(state.commandPositionWord, 0);
		this.commandSizeWord.set(state.commandSizeWord, 0);
		this.commandWordStart.set(state.commandWordStart, 0);
		this.words.set(state.words, 0);
		this.publishRevision();
	}

	public sealForPresentation(): void {
		this.presentCommandCount = this.commandCount;
	}

	public hasUnretiredPresentCommands(): boolean {
		return this.presentCommandCount !== 0;
	}

	public retirePresentedCommands(): void {
		const retiredCommands = this.presentCommandCount;
		if (retiredCommands === 0) {
			return;
		}
		const oldCommandCount = this.commandCount;
		const oldWordCount = this.wordCount;
		const retiredWords = retiredCommands < oldCommandCount
			? this.commandWordStart[retiredCommands]
			: this.activeWordsRemaining !== 0 ? this.activeWordStart : oldWordCount;
		const remainingCommands = oldCommandCount - retiredCommands;
		this.commandPositionWord.copyWithin(0, retiredCommands, oldCommandCount);
		this.commandSizeWord.copyWithin(0, retiredCommands, oldCommandCount);
		this.commandWordStart.copyWithin(0, retiredCommands, oldCommandCount);
		for (let commandIndex = 0; commandIndex < remainingCommands; commandIndex += 1) {
			this.commandWordStart[commandIndex] -= retiredWords;
		}
		this.words.copyWithin(0, retiredWords, oldWordCount);
		this.commandCount = remainingCommands;
		this.presentCommandCount = 0;
		this.wordCount = oldWordCount - retiredWords;
		if (this.activeWordsRemaining !== 0) {
			this.activeWordStart -= retiredWords;
		}
		this.publishRevision();
		this.updateStatus();
	}

	private publishRevision(): void {
		gxGpuSystemVramNextSerial = (gxGpuSystemVramNextSerial + 1) >>> 0;
		this.serial = gxGpuSystemVramNextSerial;
	}

	private abortTransfers(): void {
		this.commandCount = 0;
		this.presentCommandCount = 0;
		this.wordCount = 0;
		this.activePositionWord = 0;
		this.activeSizeWord = 0;
		this.activeWordStart = 0;
		this.activeWordsRemaining = 0;
		this.statusWord = 0;
		this.publishRevision();
	}

	private beginTransfer(): void {
		if (this.activeWordsRemaining !== 0) {
			this.statusWord |= GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW;
			this.updateStatus();
			return;
		}
		const pixelCount = gxGpuSystemVramWidth(this.sizeWord) * gxGpuSystemVramHeight(this.sizeWord);
		const transferWordCount = (pixelCount + 1) >> 1;
		if (this.commandCount === GX_GPU_SYSTEM_VRAM_PORT_COMMAND_CAPACITY
			|| this.wordCount + transferWordCount > GX_GPU_SYSTEM_VRAM_PORT_WORD_CAPACITY) {
			this.statusWord |= GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW;
			this.updateStatus();
			return;
		}
		this.activeWordStart = this.wordCount;
		this.activePositionWord = this.positionWord;
		this.activeSizeWord = this.sizeWord;
		this.activeWordsRemaining = transferWordCount;
		this.updateStatus();
	}

	private writeData(word: number): void {
		this.dataWord = word;
		if (this.activeWordsRemaining === 0) {
			return;
		}
		this.words[this.wordCount] = word;
		this.wordCount += 1;
		this.activeWordsRemaining -= 1;
		if (this.activeWordsRemaining === 0) {
			const commandIndex = this.commandCount;
			this.commandPositionWord[commandIndex] = this.activePositionWord;
			this.commandSizeWord[commandIndex] = this.activeSizeWord;
			this.commandWordStart[commandIndex] = this.activeWordStart;
			this.commandCount = commandIndex + 1;
			this.activePositionWord = 0;
			this.activeSizeWord = 0;
			this.activeWordStart = 0;
		}
		this.updateStatus();
	}

	private updateStatus(): void {
		const retainedStatus = this.statusWord & GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW;
		this.statusWord = (retainedStatus
			| (this.activeWordsRemaining !== 0
				? GX_GPU_SYSTEM_VRAM_PORT_STATUS_BUSY | GX_GPU_SYSTEM_VRAM_PORT_STATUS_WRITE_READY
				: 0)
			| (this.commandCount !== 0 ? GX_GPU_SYSTEM_VRAM_PORT_STATUS_PENDING : 0)
			| (this.activeWordsRemaining << GX_GPU_SYSTEM_VRAM_PORT_STATUS_REMAINING_SHIFT)) >>> 0;
	}

	private static readRegister(context: GxGpuSystemVramPort, address: number): number {
		switch (address) {
			case IO_GX_GPU_SYSTEM_VRAM_POSITION:
				return context.positionWord;
			case IO_GX_GPU_SYSTEM_VRAM_SIZE:
				return context.sizeWord;
			case IO_GX_GPU_SYSTEM_VRAM_CONTROL:
				return context.controlWord;
			case IO_GX_GPU_SYSTEM_VRAM_DATA:
				return context.dataWord;
			case IO_GX_GPU_SYSTEM_VRAM_STATUS:
				return context.statusWord;
		}
		throw gxGpuSystemVramUnmappedReadError;
	}

	private static writeRegister(context: GxGpuSystemVramPort, address: number, value: Value): void {
		const word = (value as number) >>> 0;
		switch (address) {
			case IO_GX_GPU_SYSTEM_VRAM_POSITION:
				context.positionWord = word;
				return;
			case IO_GX_GPU_SYSTEM_VRAM_SIZE:
				context.sizeWord = word;
				return;
			case IO_GX_GPU_SYSTEM_VRAM_CONTROL:
				context.controlWord = word;
				if ((word & GX_GPU_SYSTEM_VRAM_PORT_CONTROL_RESET) !== 0) {
					context.abortTransfers();
				}
				if ((word & GX_GPU_SYSTEM_VRAM_PORT_CONTROL_START) !== 0) {
					context.beginTransfer();
				}
				return;
			case IO_GX_GPU_SYSTEM_VRAM_DATA:
				context.writeData(word);
				return;
		}
		throw gxGpuSystemVramUnmappedWriteError;
	}
}

File: machine/cpp/machine/devices/gx/system_vram_port.cpp
#include "machine/devices/gx/system_vram_port.h"

#include "machine/bus/io.h"
#include "machine/cpu/cpu.h"
#include "machine/devices/gx/gpu_command_buffer.h"
#include "machine/memory/memory.h"

#include <algorithm>

namespace bmsx {

GxGpuSystemVramPort::GxGpuSystemVramPort(Memory& memory) {
	memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_POSITION, this, &GxGpuSystemVramPort::readRegister);
	memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_SIZE, this, &GxGpuSystemVramPort::readRegister);
	memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_CONTROL, this, &GxGpuSystemVramPort::readRegister);
	memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_DATA, this, &GxGpuSystemVramPort::readRegister);
	memory.mapIoRead(IO_GX_GPU_SYSTEM_VRAM_STATUS, this, &GxGpuSystemVramPort::readRegister);
	memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_POSITION, this, &GxGpuSystemVramPort::writeRegister);
	memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_SIZE, this, &GxGpuSystemVramPort::writeRegister);
	memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_CONTROL, this, &GxGpuSystemVramPort::writeRegister);
	memory.mapIoWrite(IO_GX_GPU_SYSTEM_VRAM_DATA, this, &GxGpuSystemVramPort::writeRegister);
}

void GxGpuSystemVramPort::reset() {
	m_positionWord = 0u;
	m_sizeWord = 0u;
	m_controlWord = 0u;
	m_dataWord = 0u;
	m_statusWord = 0u;
	commandCount = 0u;
	presentCommandCount = 0u;
	wordCount = 0u;
	m_activePositionWord = 0u;
	m_activeSizeWord = 0u;
	m_activeWordStart = 0u;
	m_activeWordsRemaining = 0u;
	publishRevision();
}

GxGpuSystemVramPortState GxGpuSystemVramPort::captureState() const {
	GxGpuSystemVramPortState state;
	state.positionWord = m_positionWord;
	state.sizeWord = m_sizeWord;
	state.controlWord = m_controlWord;
	state.dataWord = m_dataWord;
	state.statusWord = m_statusWord;
	state.commandCount = commandCount;
	state.presentCommandCount = presentCommandCount;
	state.wordCount = wordCount;
	state.activePositionWord = m_activePositionWord;
	state.activeSizeWord = m_activeSizeWord;
	state.activeWordStart = m_activeWordStart;
	state.activeWordsRemaining = m_activeWordsRemaining;
	state.commandPositionWord.assign(commandPositionWord.begin(), commandPositionWord.begin() + static_cast<std::ptrdiff_t>(commandCount));
	state.commandSizeWord.assign(commandSizeWord.begin(), commandSizeWord.begin() + static_cast<std::ptrdiff_t>(commandCount));
	state.commandWordStart.assign(commandWordStart.begin(), commandWordStart.begin() + static_cast<std::ptrdiff_t>(commandCount));
	state.words.assign(words.begin(), words.begin() + static_cast<std::ptrdiff_t>(wordCount));
	return state;
}

void GxGpuSystemVramPort::restoreState(const GxGpuSystemVramPortState& state) {
	m_positionWord = state.positionWord;
	m_sizeWord = state.sizeWord;
	m_controlWord = state.controlWord;
	m_dataWord = state.dataWord;
	m_statusWord = state.statusWord;
	commandCount = state.commandCount;
	presentCommandCount = state.presentCommandCount;
	wordCount = state.wordCount;
	m_activePositionWord = state.activePositionWord;
	m_activeSizeWord = state.activeSizeWord;
	m_activeWordStart = state.activeWordStart;
	m_activeWordsRemaining = state.activeWordsRemaining;
	std::copy(state.commandPositionWord.begin(), state.commandPositionWord.end(), commandPositionWord.begin());
	std::copy(state.commandSizeWord.begin(), state.commandSizeWord.end(), commandSizeWord.begin());
	std::copy(state.commandWordStart.begin(), state.commandWordStart.end(), commandWordStart.begin());
	std::copy(state.words.begin(), state.words.end(), words.begin());
	publishRevision();
}

void GxGpuSystemVramPort::sealForPresentation() {
	presentCommandCount = commandCount;
}

void GxGpuSystemVramPort::retirePresentedCommands() {
	const size_t retiredCommands = presentCommandCount;
	if (retiredCommands == 0u) {
		return;
	}
	const size_t oldCommandCount = commandCount;
	const size_t oldWordCount = wordCount;
	const size_t retiredWords = retiredCommands < oldCommandCount
		? commandWordStart[retiredCommands]
		: m_activeWordsRemaining != 0u ? m_activeWordStart : oldWordCount;
	const size_t remainingCommands = oldCommandCount - retiredCommands;
	std::move(commandPositionWord.begin() + static_cast<std::ptrdiff_t>(retiredCommands), commandPositionWord.begin() + static_cast<std::ptrdiff_t>(oldCommandCount), commandPositionWord.begin());
	std::move(commandSizeWord.begin() + static_cast<std::ptrdiff_t>(retiredCommands), commandSizeWord.begin() + static_cast<std::ptrdiff_t>(oldCommandCount), commandSizeWord.begin());
	std::move(commandWordStart.begin() + static_cast<std::ptrdiff_t>(retiredCommands), commandWordStart.begin() + static_cast<std::ptrdiff_t>(oldCommandCount), commandWordStart.begin());
	for (size_t commandIndex = 0u; commandIndex < remainingCommands; commandIndex += 1u) {
		commandWordStart[commandIndex] -= static_cast<u32>(retiredWords);
	}
	std::move(words.begin() + static_cast<std::ptrdiff_t>(retiredWords), words.begin() + static_cast<std::ptrdiff_t>(oldWordCount), words.begin());
	commandCount = remainingCommands;
	presentCommandCount = 0u;
	wordCount = oldWordCount - retiredWords;
	if (m_activeWordsRemaining != 0u) {
		m_activeWordStart -= retiredWords;
	}
	publishRevision();
	updateStatus();
}

void GxGpuSystemVramPort::publishRevision() {
	nextSerial += 1u;
	serial = nextSerial;
}

void GxGpuSystemVramPort::abortTransfers() {
	commandCount = 0u;
	presentCommandCount = 0u;
	wordCount = 0u;
	m_activePositionWord = 0u;
	m_activeSizeWord = 0u;
	m_activeWordStart = 0u;
	m_activeWordsRemaining = 0u;
	m_statusWord = 0u;
	publishRevision();
}

void GxGpuSystemVramPort::beginTransfer() {
	if (m_activeWordsRemaining != 0u) {
		m_statusWord |= GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW;
		updateStatus();
		return;
	}
	const u32 pixelCount = gxGpuSystemVramWidth(m_sizeWord) * gxGpuSystemVramHeight(m_sizeWord);
	const u32 transferWordCount = (pixelCount + 1u) >> 1u;
	if (commandCount == GX_GPU_SYSTEM_VRAM_PORT_COMMAND_CAPACITY
		|| wordCount + transferWordCount > GX_GPU_SYSTEM_VRAM_PORT_WORD_CAPACITY) {
		m_statusWord |= GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW;
		updateStatus();
		return;
	}
	m_activeWordStart = wordCount;
	m_activePositionWord = m_positionWord;
	m_activeSizeWord = m_sizeWord;
	m_activeWordsRemaining = transferWordCount;
	updateStatus();
}

void GxGpuSystemVramPort::writeData(u32 word) {
	m_dataWord = word;
	if (m_activeWordsRemaining == 0u) {
		return;
	}
	words[wordCount] = word;
	wordCount += 1u;
	m_activeWordsRemaining -= 1u;
	if (m_activeWordsRemaining == 0u) {
		const size_t commandIndex = commandCount;
		commandPositionWord[commandIndex] = m_activePositionWord;
		commandSizeWord[commandIndex] = m_activeSizeWord;
		commandWordStart[commandIndex] = static_cast<u32>(m_activeWordStart);
		commandCount = commandIndex + 1u;
		m_activePositionWord = 0u;
		m_activeSizeWord = 0u;
		m_activeWordStart = 0u;
	}
	updateStatus();
}

void GxGpuSystemVramPort::updateStatus() {
	const u32 retainedStatus = m_statusWord & GX_GPU_SYSTEM_VRAM_PORT_STATUS_OVERFLOW;
	m_statusWord = retainedStatus
		| (m_activeWordsRemaining != 0u ? GX_GPU_SYSTEM_VRAM_PORT_STATUS_BUSY | GX_GPU_SYSTEM_VRAM_PORT_STATUS_WRITE_READY : 0u)
		| (commandCount != 0u ? GX_GPU_SYSTEM_VRAM_PORT_STATUS_PENDING : 0u)
		| (m_activeWordsRemaining << GX_GPU_SYSTEM_VRAM_PORT_STATUS_REMAINING_SHIFT);
}

u64 GxGpuSystemVramPort::readRegister(void* context, u32 address) {
	auto& port = *static_cast<GxGpuSystemVramPort*>(context);
	switch (address) {
	case IO_GX_GPU_SYSTEM_VRAM_POSITION:
		return valueNumber(static_cast<double>(port.m_positionWord));
	case IO_GX_GPU_SYSTEM_VRAM_SIZE:
		return valueNumber(static_cast<double>(port.m_sizeWord));
	case IO_GX_GPU_SYSTEM_VRAM_CONTROL:
		return valueNumber(static_cast<double>(port.m_controlWord));
	case IO_GX_GPU_SYSTEM_VRAM_DATA:
		return valueNumber(static_cast<double>(port.m_dataWord));
	case IO_GX_GPU_SYSTEM_VRAM_STATUS:
		return valueNumber(static_cast<double>(port.m_statusWord));
	}
	throw BMSX_RUNTIME_ERROR("GX system VRAM port read outside mapped registerfile.");
}

void GxGpuSystemVramPort::writeRegister(void* context, u32 address, u64 value) {
	auto& port = *static_cast<GxGpuSystemVramPort*>(context);
	const u32 word = toU32(value);
	switch (address) {
	case IO_GX_GPU_SYSTEM_VRAM_POSITION:
		port.m_positionWord = word;
		return;
	case IO_GX_GPU_SYSTEM_VRAM_SIZE:
		port.m_sizeWord = word;
		return;
	case IO_GX_GPU_SYSTEM_VRAM_CONTROL:
		port.m_controlWord = word;
		if ((word & GX_GPU_SYSTEM_VRAM_PORT_CONTROL_RESET) != 0u) {
			port.abortTransfers();
		}
		if ((word & GX_GPU_SYSTEM_VRAM_PORT_CONTROL_START) != 0u) {
			port.beginTransfer();
		}
		return;
	case IO_GX_GPU_SYSTEM_VRAM_DATA:
		port.writeData(word);
		return;
	}
	throw BMSX_RUNTIME_ERROR("GX system VRAM port write outside mapped registerfile.");
}

} // namespace bmsx
